examples/kalman_filter.py

```diff
@@ -0,0 +1,374 @@
+
+        Here the linear error observation - expected is brought back to
+        the Lie algebra
+        """
+        theta, _, _ = state
+        rot = self.rotation_matrix(theta)
+        expected = Localization.observation_model(state)
+        return rot.T.dot(observation - expected)
+
+
+class KalmanFilter:
+    """Class for a general Kalman filter working on Lie groups.
+
+    Given an adapted model, it provides the tools to carry out non-linear state
+    estimation with an error modeled on the Lie algebra.
+    """
+
+    def __init__(self, model):
+        self.model = model
+        self.state = model.group.get_identity()
+        self.covariance = gs.zeros((self.model.dim, self.model.dim))
+        self.process_noise = gs.zeros(
+            (self.model.dim_noise, self.model.dim_noise))
+        self.measurement_noise = gs.zeros(
+            (self.model.dim_obs, self.model.dim_obs))
+
+    def initialise_covariances(self, prior_values, process_values, obs_values):
+        """Set the values of the covariances."""
+        values = [prior_values, process_values, obs_values]
+        attributes = ['covariance', 'process_noise', 'measurement_noise']
+        for (index, val) in enumerate(values):
+            if gs.ndim(val) == 1:
+                setattr(self, attributes[index],
+                        algebra_utils.from_vector_to_diagonal_matrix(val))
+            else:
+                setattr(self, attributes[index], val)
+
+    def propagate(self, sensor_input):
+        """Propagate the estimate and its covariance."""
+        prop_noise = self.process_noise
+        prop_jac = self.model.propagation_jacobian(self.state, sensor_input)
+        noise_jac = self.model.noise_jacobian(self.state, sensor_input)
+
+        self.covariance = prop_jac.dot(self.covariance).dot(
+            prop_jac.T) + noise_jac.dot(prop_noise).dot(noise_jac.T)
+        self.state = self.model.propagate(self.state, sensor_input)
+
+    def compute_gain(self, observation):
+        """Compute the Kalman gain given the observation model."""
+        N = self.model.get_measurement_noise_cov(
+            self.state, self.measurement_noise)
+        obs_jac = self.model.observation_jacobian(self.state, observation)
+        innovation_cov = obs_jac.dot(self.covariance).dot(obs_jac.T) + N
+        return self.covariance.dot(obs_jac.T).dot(
+            gs.linalg.inv(innovation_cov))
+
+    def update(self, observation):
+        """Update the current estimate given an observation."""
+        innovation = self.model.innovation(self.state, observation)
+        gain = self.compute_gain(observation)
+        obs_jac = self.model.observation_jacobian(self.state, observation)
+        self.covariance = (gs.eye(self.model.dim) - gain.dot(obs_jac)).dot(
+            self.covariance)
+        self.state = self.model.group.exp(gain.dot(innovation), self.state)
+        if gs.ndim(self.state) > 1:
+            return gs.squeeze(self.state)
+
+
+def main():
```

Comment:
It would be great to have this example in a python notebook.

Reply:
Should everything be in the notebook ? Or only the part after the classes are defined ?

Reply:
I would say only the parts after the class @ninamiolane thoughts?

Reply:
Agreed!